### PR TITLE
Handle arrays of empty elements properly

### PR DESCRIPTION
**Pull request by Mange**
It used to be that arrays containing empty string members returned their members as stringified hashes of the attributes instead of the empty string:

```xml
  <item>
    <foo xsi:type="xsd:string"></foo>
  </item>
  <item>
    <foo xsi:type="xsd:string"></foo>
  </item>
```

```ruby
  expect(list.first["foo"]).to eq("")
  # Is actually: '{"@xsi:type"=>"xsd:string"}`
```

This is a pretty conservative change as it only kicks in when:
1.  A parameter that shouldn't be a `struct` is a `Hash`.
2. All the keys in that hash are *attributes*.

That means that if the element was actually nested elements, it will still be returned as a stringified Hash.

---

I had a hard time making this understandable as this is an effect of how Nori is converting XML back and forth. Should I add the example XML to a comment to make it easier to understand, or is this enough?

### CODE REVIEW

**Comment by rmeritz** [comment]
This solution works which is great. But I already worked on fixing a similar bug and fixed in a completely different part of the system. Wouldn't it make more sense to consolidate this fixes?

https://github.com/inossidabile/wash_out/pull/196

**Comment by Mange** [comment]
> But I already worked on fixing a similar bug and fixed in a completely different part of the system. Wouldn't it make more sense to consolidate this fixes?

Thanks for linking me to that!

I'm not sure the issues themselves are related, though. They both have the same origin[1], but these two sections of the code does not appear to work on the same problems as far as I can see.

That code is handling references and the data structure itself when it comes to matching it to the type definitions, but the code in this PR is doing all the type conversions after that structure is determined.

Maybe @inossidabile can clarify where this fix should end up? :-)

[1]: This class of problems comes `wash_out` using `Nori` for XML parsing, and the fact that a highly complicated data structure is reduced down to Ruby primitives – losing details and structure in the process.

**Comment by inossidabile** [comment]
Well... I disagree that the problem is that highly complicated structure is reduced. In fact it's just that Nori is incredibly stupid. I also don't really see how it's related to #196 at all. Can you spread the light on me @rmeritz?

**Comment by rmeritz** [comment]
@inossidabile - I think I was looking at the symptom (which exactly the same) instead of the problem itself so I withdraw my prior comment.

**Comment by inossidabile** [comment]
So I investigated and have two things to say:

1. Your fix is incorrect because it only affects `Param` children. That won't work with the general parses and this can't be merged.
2. I tried to do what you asked for but it seems that it can't be done well with Nori. Moreover I digged the history of this issue and it seems that ruby community got quite used to this weird behavior since it comes with Savon and Savon is the SOAP thing of ours. So in fact Nori started with proposing you to perceive every node (even nilable) as either a hash or an array. It even subclassed the string to maintain hash nature for text nodes. But later it all got fucked up by advanced typecasting...

In other words - if you want it fixed - you need to send patch to Nori. I will not merge the workaround in WashOut, sorry.
